std/typecons.d

```diff
@@ -6896,3 +6897,503 @@ unittest
+        version(assert)
+        {
+            import core.exception : RangeError;
+            if (!isDefined || index != 0) throw new RangeError();
+        }
+        return value;
+    }
+
+    /// Ditto
+    Option opSlice(size_t from, size_t to)
+    {
+        version(assert)
+        {
+            import core.exception : RangeError;
+            if (from > to || length < to) throw new RangeError();
+        }
+        return to > from ? this : Option(none);
+    }
+
+    /**
+     * Assume this instance holds a value and return it.
+     *
+     * Range algorithms are preferable over raw access to the underlying
+     * value when possible. In other cases, using Option and assumeNonEmpty is
+     * preferable to using a raw nullable reference, as the `AssertError` is
+     * guaranteed to be thrown, and unlike a segmentation fault or other
+     * undefined behavior, will be as close to the underlying error as
+     * possible. Further, using assumeNonEmpty formalizes the assumption of
+     * non-null.
+     * Returns: The held value
+     * Throws: `AssertError` if this instance is empty
+     * Note: `opt.assumeNonEmpty = null;` does not empty the instance; to empty
+     * the instance, assign to $(LREF none).
+     * Bugs: This function is currently not `nothrow` or `@nogc`.
+     * $(LREF Option.front) can be used as a workaround when these attributes
+     * are desired. See $(BUGZILLA 12647).
+     */
+    // this function is not nothrow due to bug 12647, and not @nogc.
+    // onAssertError isn't @nogc either.
+    ref inout(T) assumeNonEmpty() inout @property pure @safe
+    {
+        import core.exception : AssertError;
+        import std.exception : enforce;
+        enforce!AssertError(isDefined, T.stringof ~ " asssumption failed to hold");
+        return value;
+    }
+
+    /**
+     * If there is a held value, return it; otherwise return other.
+     * Params: other = value to return if this instance is empty
+     * Bugs: These functions are currently not `nothrow` or `@nogc`. See
+     * $(BUGZILLA 12647).
+     * See_Also: $(REF Algebraic, std, variant)
+     */
+    // TODO: Do lazy parameters offer ref yet?
+    auto ref CommonType!(inout(T), U) getOrElse(U : T)(lazy U other) inout
+    {
+        return isDefined ? value : other;
+    }
+
+    /// Ditto
+    Option!(CommonType!(T, U)) orElse(U)(lazy Option!U other)
+        if (!is(CommonType!(T, U) == void))
+    {
+        alias Super = CommonType!(T, U);
+        return isDefined ? some!Super(this.value) : other.upcast!Super;
+    }
+
+    version(StdDdoc)
+    {
+        /// Ditto
+        Algebraic!(inout(T), U) or(lazy U other) inout;
+    }
+    else
+    {
+        template or(U)
+        {
+            import std.variant : Algebraic;
+            Algebraic!(T, U) or(lazy U other)
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+
+            Algebraic!(const T, U) or(lazy U other) const
```

Comment:
`const` and `immutable` overloads are untested because `Algebraic` does not support non-mutable types...